### DISCUSSION TITLE
Accessing package data

### DISCUSSION THREAD

**Discussion by bicarlsen** [PyO3/maturin#1543] [Q&A] [answered]
I have included some package data as [described in the guide](https://www.maturin.rs/project_layout.html#data), but am unsure the best way to get its path once installed. In Python I would use [`pkg_resources.resource_filename`](https://setuptools.pypa.io/en/latest/pkg_resources.html#resource-extraction). What is the equivalent for `maturin`?

**Comment by messense** [comment]
I think it's the same for maturin since maturin does nothing special. If it doesn't work for you, it might be a implementation bug in maturin, in that case please file a bug report.

cc @konstin

**Reply by bicarlsen** [reply]
Ah, sorry, I meant how do I get this path in the Rust code.

**Reply by messense** [reply]
Well, you can also import `pkg_resources` in Rust, for example

```rust
#[pyfunction]
fn use_pkg_resources(py: Python<'_>) -> PyResult<()> {
    let pkg_resources = py.import("pkg_resources")?;
    let resource_filename = pkg_resources.getattr("resource_filename")?;
    let resource = resource_filename.call(("my_module_name", "my_resource_name"), None)?;
    todo!()
}

```

**Reply by messense** [reply]
But if you only need to access the resources in Rust code, I'd suggest using [`include_bytes!`](https://doc.rust-lang.org/std/macro.include_bytes.html) or [`include_str!`](https://doc.rust-lang.org/std/macro.include_str.html) instead.

**Comment by bicarlsen** [comment]
Based on @messense response I came up with this to get the base path to a package.

```rust
fn resources_path(py: Python<'_>, package: &str) -> PyResult<PathBuf> {
    let resources = py.import("importlib.resources")?;
    let files = resources.call_method1("files", (package,))?;
    let files = resources.call_method1("as_file", (files,))?;
    let path = files.call_method0("__enter__")?; // enter python context manager
    let path: PathBuf = path.extract()?;
    let none = py.None();
    files.call_method1("__exit__", (&none, &none, &none))?; // exit python context manager

    Ok(path)
}
```
If anybody has feedback on the "goodness" of this, I would greatly appreciate the code review.

**Reply by bicarlsen** [reply, marked as answer]
Perhaps I should call `__exit__` based on the result of `path.extract`?
```rust
fn resources_path(py: Python<'_>, package: &str) -> PyResult<PathBuf> {
    let resources = py.import("importlib.resources")?;
    let files = resources.call_method1("files", (package,))?;
    let files = resources.call_method1("as_file", (files,))?;
    let path = files.call_method0("__enter__")?; // enter python context manager
    match path.extract() {
        Ok(path) => {
            let none = py.None();
            files.call_method1("__exit__", (&none, &none, &none))?;
            Ok(path)
        }, 
        Err(err) => {
            files.call_method1(
                "__exit__",
                (err.get_type(py), err.value(py), err.traceback(py))
            ).unwrap();

            Err(err)
        }
    }
}
```

**Reply by messense** [reply]
See https://pyo3.rs/latest/python_from_rust#need-to-use-a-context-manager-from-rust

**Reply by bicarlsen** [reply]
Thank you. This is what I based my code off of, but wasn't sure the best way to implement in this use case. I decided to go with the version calling `exit` based on the results of `path.extract`.

**Comment by bicarlsen** [comment]
While this seems to get the path to the `site_packages/<my_package>` folder, nothing from the `<my_package>.data/data` folder appears to be there.

**Reply by messense** [reply]
If it's unexpected, please file a bug report with a repro, thanks!

**Reply by bicarlsen** [reply]
I've raised Issue #1548 . Thank you for all the help :)
